src/rewards/ca/mod.rs

```diff
@@ -0,0 +1,201 @@
+pub type CaRewards = HashMap<Ca, Rewards>;
+pub type ProposalsReviews = HashMap<ProposalId, Vec<AdvisorReviewRow>>;
+pub type ApprovedProposals = HashSet<ProposalId>;
+
+enum ProposalFundsState {
+    // Proposal has the exact quantity reviews to be rewarded
+    Exact,
+    // Proposal has less reviews as needed so some of the funds should go back into the rewards pool
+    Unfilled(Funds),
+    // It has more reviews than fitted rewards
+    OverLoaded,
+}
+
+struct ProposalReward {
+    pub state: ProposalFundsState,
+    pub funds: Funds,
+}
+
+fn proposal_rewards_state(
+    proposal_reviews: &[AdvisorReviewRow],
+    proposal_fund: Funds,
+    rewards_slots: &ProposalRewardSlots,
+) -> ProposalFundsState {
+    let filled_slots: u64 = proposal_reviews
+        .iter()
+        .map(|review| match review.score() {
+            ReviewScore::Excellent => rewards_slots.excellent_slots,
+            ReviewScore::Good => rewards_slots.good_slots,
+        })
+        .sum();
+
+    match filled_slots.cmp(&rewards_slots.filled_slots) {
+        Ordering::Less => {
+            let unfilled_funds = proposal_fund
+                * (Funds::from(filled_slots) / Funds::from(rewards_slots.filled_slots));
+            ProposalFundsState::Unfilled(unfilled_funds)
+        }
+        Ordering::Equal => ProposalFundsState::Exact,
+
+        Ordering::Greater => ProposalFundsState::OverLoaded,
+    }
+}
+
+fn calculate_funds_per_proposal(
+    proposal_reviews: &ProposalsReviews,
+    approved_proposals: &ApprovedProposals,
+    funding: &FundSetting,
+    rewards_slots: &ProposalRewardSlots,
+) -> ProposalsFunds {
+    let per_proposal_reward = funding.funds_per_proposal(proposal_reviews.len() as u64);
+    let bonus_proposals_rewards = funding.bonus_funds_per_proposal(approved_proposals.len() as u64);
+
+    // check rewards and split extra until there is no more to split
+    let proposal_rewards_states: HashMap<ProposalId, ProposalFundsState> = proposal_reviews
+        .iter()
+        .map(|(id, reviews)| {
+            (
+                id.clone(),
+                proposal_rewards_state(reviews, per_proposal_reward, rewards_slots),
+            )
+        })
+        .collect();
+
+    let underbudget_funds: Funds = proposal_rewards_states
+        .values()
+        .map(|state| match state {
+            ProposalFundsState::Unfilled(value) => *value,
+            _ => Funds::from(0u64),
+        })
+        .sum();
+
+    let underbudget_rewards = underbudget_funds / Funds::from(proposal_reviews.len() as u64);
+
+    proposal_rewards_states
+        .into_iter()
+        .map(|(id, state)| {
+            let bonus_funds = approved_proposals
+                .contains(&id)
+                .then(|| bonus_proposals_rewards)
+                .unwrap_or_else(|| Funds::from(0u64));
+            let funds = match state {
+                ProposalFundsState::Unfilled(unfilled_funds) => {
+                    per_proposal_reward - unfilled_funds
```

Comment:
I think `underbudget_rewards` should be added to unfilled proposals as well

Reply:
Im not sure ablut that. But it is a point of discussion because then it would have more budget again... 
But, it also makes sense to make just another iteration like that. Let me ask Marek about this.